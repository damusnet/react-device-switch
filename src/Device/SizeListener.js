import React, { createContext, Component } from "react";
import PropTypes from "prop-types";

export const MOBILE = "MOBILE";
export const DESKTOP = "DESKTOP";

export const DeviceTypeContext = createContext(MOBILE);

class DeviceSizeListener extends Component {
  constructor(props) {
    super(props);
    this.state = { deviceType: MOBILE };
    this.resize = this.resize.bind(this);
  }

  componentDidMount() {
    const { breakpoint } = this.props;
    if (typeof window !== "undefined" && window.matchMedia) {
      this.media = window.matchMedia(`(${breakpoint})`);
      this.media.addListener(this.resize);
      this.resize();
    }
  }

  componentWillUnmount() {
    if (this.media) {
      this.media.removeListener(this.resize);
    }
  }

  resize() {
    const { deviceType } = this.state;

    const realDeviceType = this.media.matches ? DESKTOP : MOBILE;

    if (deviceType !== realDeviceType) {
      this.setState({ deviceType: realDeviceType });
    }
  }

  render() {
    const { children } = this.props;
    const { deviceType } = this.state;
    return (
      <DeviceTypeContext.Provider value={deviceType}>
        {children}
      </DeviceTypeContext.Provider>
    );
  }
}

DeviceSizeListener.propTypes = {
  breakpoint: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired
};

export default DeviceSizeListener;
