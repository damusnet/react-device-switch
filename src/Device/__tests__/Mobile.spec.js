import React from "react";
import { render, cleanup } from "react-testing-library";

import { DeviceTypeContext, MOBILE, DESKTOP } from "../SizeListener";
import DeviceMobile from "../Mobile";

describe("<DeviceMobile />", () => {
  afterEach(cleanup);

  it("renders mobile content correctly on MOBILE", () => {
    const { container } = render(
      <DeviceTypeContext.Provider value={MOBILE}>
        <DeviceMobile>Mobile Content</DeviceMobile>
      </DeviceTypeContext.Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders no mobile content on DESKTOP", () => {
    const { container } = render(
      <DeviceTypeContext.Provider value={DESKTOP}>
        <DeviceMobile>Mobile Content</DeviceMobile>
      </DeviceTypeContext.Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
