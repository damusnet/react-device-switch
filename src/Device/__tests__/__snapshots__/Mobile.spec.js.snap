// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeviceMobile /> renders mobile content correctly on MOBILE 1`] = `Mobile Content`;

exports[`<DeviceMobile /> renders no mobile content on DESKTOP 1`] = `null`;
